Quadlet: ensure all keys are documented

edsantiago opened the pull request:
New CI validation check: all keys in quadlet.go must be
documented at least once in podman-systemd.unit.5.md.

And, because the md file lists keys in both table and block
form, make sure those all match.

And make sure everything is sorted in lexical order.

You'd think that by now I would know enough to add
[NO NEW TESTS NEEDED]

Signed-off-by: Ed Santiago <santiago@redhat.com>
```release-note
None
```

edsantiago commented on docs/source/markdown/podman-systemd.unit.5.md:
```diff
@@ -204,8 +203,14 @@ Valid options for `[Container]` are listed below:
 | Pull=never                           | --pull=never                                         |
 | ReadOnly=true                        | --read-only                                          |
 | ReadOnlyTmpfs=true                   | --read-only-tmpfs                                    |
+| RemapGid=FIXME                       | FIXME                                                |
+| RemapUid=FIXME                       | FIXME                                                |
+| RemapUidSize=FIXME                   | FIXME                                                |
+| RemapUsers=FIXME                     | FIXME                                                |
```
Someone please tell me how to fill these in, and if they're in the right unit section

ygalblum replied:
These keys were deprecated and therefore remove from the documentation. They are still supported in the code to maintain backward compatibility.

edsantiago commented on hack/xref-quadlet-docs:
```diff
@@ -0,0 +1,261 @@
+            my ($constname, $keystring) = ($1, $2);
+
+            # const name *should* be the same as the string, except (sigh)
+            # there's an inexplicable block of KeyNetworkFoo = "Foo".
```
"inexplicable" to me. If there's a great reason for it, please lmk so I can update the comment.

ygalblum replied:
When I added these keys the idea was to indicate that they relate the `.network` files (and hence used in `ConvertNetwork`). I don't see any harm in removing `Network` if you want

edsantiago commented on hack/xref-quadlet-docs:
```diff
@@ -0,0 +1,261 @@
+            # const name *should* be the same as the string, except (sigh)
+            # there's an inexplicable block of KeyNetworkFoo = "Foo".
+            unless ($constname eq $keystring || $constname eq "Network$keystring") {
+                # There's also, sigh again, some preexisting mismatches
+                unless ($constname =~ /Remap[GU]ID/) {
```
Barf. Since these are all undocumented, would it be possible to actually fix them?

ygalblum replied:
As I explained, these keys were deprected. I see that the code looks for `deprecated`. So, I guess the solution is to add this comment to these lines

edsantiago commented on hack/xref-quadlet-docs:
```diff
@@ -0,0 +1,261 @@
+            # Dup check. Assume that Go compiler will detect duplicate
+            # const names, but we have to check the strings.
+            if (grep { lc($_) eq lc($keystring) } @found) {
+                # sigh: "Driver" and "Options" also exist in KeyNetworkFoo space
+                warn "$ME: $path:$.: duplicate key string \"$keystring\"\n"
+                    unless $keystring =~ /^(Driver|Options)$/;
```
Barf again. Just checking: is it intentional to have multiple `Driver` and `Options` definitions?

Luap99 replied:
I think that is something that should be fixed in the code.
I guess this is due the fact that we have Driver/Options for Volume and Network units for example. So it would need to be documented for each unit type. But yeah there is really no reason to define the same string const twice in the code unless we like to bloat the binary size by some bytes.

ygalblum replied:
As I wrote, the `Network` part can be removed (and then also remove the duplications)

edsantiago reviewed (commented):
Blocked pending resolution of many questions.

Luap99 commented on docs/source/markdown/podman-systemd.unit.5.md:
```diff
@@ -690,6 +712,10 @@ which can be modified with `UserNS`, but if that is not specified, this UID is a
 Set the user namespace mode for the container. This is equivalent to the Podman `--userns` option and
 generally has the form `MODE[:OPTIONS,...]`.
 
+### `VolatileTmp=`
+
+FIXME
```
I would need to look through the git log but if I remember correctly this key is deprecated and thus not documented.
I like the idea of this but I think we need to have a way to ignore some keys. Maybe we could add `// deprecated` as comment in the code behind the key name and you script could see it and not force the docs or something like this.

edsantiago replied:
Thank you! #20479. I had missed that. And I like the idea of `// deprecated`.

edsantiago commented:
Thank you @ygalblum. I've removed the `Network` prefix, fixed the case mismatch in `RemapXID`, and removed those special-case exceptions from the script. For ease of review, I've pushed this as two commits.

ygalblum commented on docs/source/markdown/podman-systemd.unit.5.md:
```diff
@@ -835,6 +836,7 @@ Valid options for `[Kube]` are listed below:
 | AutoUpdate=registry                 | --annotation "io.containers.autoupdate=registry"                 |
 | ConfigMap=/tmp/config.map           | --config-map /tmp/config.map                                     |
 | ContainersConfModule=/etc/nvd\.conf | --module=/etc/nvd\.conf                                          |
+| ExitCodePropagation=how             | How to propagate container error status              |
```
Please align `|` to the end of the table

edsantiago replied:
ARGH! Thank you for catching that. Fixed, but I will not push until I see how CI does.

ygalblum reviewed (approved):
LGTM

edsantiago commented:
rebased out of paranoia. This is ready to go.

rhatdan commented:
/lgtm
